### PR TITLE
integrate cookie-signature

### PR DESCRIPTION
**Pull request by Uzlopak**
I know you maybe dont like to integrate the cookie-signature package, but I think it has some benefits.

- We can specify which algorithm we want to use for the signing. So you can specify e.g. sha512 or sha1 instead of sha256
- standalone unsign-function will be able to handle multiple secrets and will return an UnsignResult Object, basically unifying the behaviour of unsign from the signerFactory with the standalone unsign-Function 
- because the return value of the unsign function is changed, it probably means a semver major

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: signer.js [hunk 1]
```diff
@@ -1,32 +1,91 @@
...
 
-      return { valid, renew, value }
+      return {
+        valid: false,
+        renew: false,
+        value: null
+      }
+    }
+  }
+}
```
Comment (jsumners) [inline signer.js:63]: This is way too much code inside of an object literal. The methods should be extracted into function declarations.

**Review by jsumners — CHANGES_REQUESTED** [review]
Have we tried to get the feature added upstream instead of copying their code here?

File: signer.js [hunk 2]
```diff
@@ -1,32 +1,91 @@
...
+
+// create a signer-instance, with dummy secret
+const signer = signerFactory(['dummy'])
```
Comment (mcollina) [inline signer.js:66]: This seems really unsafe.

File: signer.js [hunk 3]
```diff
@@ -1,32 +1,91 @@
...
+    sign (value, secret = signingKey, algorithm = factoryAlgorithm) {
+      if (typeof value !== 'string') {
+        throw new TypeError('Cookie value must be provided as a string.')
+      }
+      return value + '.' + crypto
+        .createHmac(algorithm, secret)
+        .update(value)
+        .digest('base64')
+        // remove base64 padding (=) as it has special meaning in cookies
+        .replace(/=+$/, '')
```
Comment (mcollina) [inline signer.js:33]: Please extract the cookie-signature bits into its own file and include the relative license disclaimer in there.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: signer.js (hunk 2 above)
Comment (Uzlopak) [reply signer.js:66]: It is just to pass the secret validation. The secret itself will be validated again in the corresponding sign and unsign function

File: signer.js (hunk 1 above)
Comment (Uzlopak) [reply signer.js:63]: Is this now better?

File: signer.js [hunk 4]
```diff
@@ -1,32 +1,96 @@
...
+
+  return {
+    valid: false,
+    renew: false,
+    value: null
+  }
+}
+
+// create a signer-instance, with dummy secret
+const signer = new SignerFactory(['dummy'])
```
Comment (mcollina) [inline signer.js:76]: If I'm reading this code correctly, `dummy` would be used for `sign` if no secret is passed in the module-level `sign` function. This is unsafe, we should not have a default secret like this.

File: signer.js [hunk 5]
```diff
@@ -1,32 +1,96 @@
...
-const cookieSignature = require('cookie-signature')
+const crypto = require('crypto')
 
-module.exports = function (secret) {
-  const secrets = Array.isArray(secret) ? secret : [secret]
-  const [signingKey] = secrets
+function SignerFactory (secrets, algorithm = 'sha256') {
+  if (!(this instanceof SignerFactory)) {
+    return new SignerFactory(secrets, algorithm)
+  }
```
Comment (jsumners) [inline signer.js:8]: This is incorrect. In JavaScript land, this is no longer a "factory" function but a "class function". So the export would either be `function Signer(...) {...}` or the original factory would return a new instance of `Signer`.

My original feedback was asking for:

```js
module.exports = function () {
  // do some stuff

  return { sign, unsign }

  function sign() {
    // sign code
  }

  function unsign() {
    // unsign code
  }
}
```

I'm not opposed have a `Signer` "class" that the exported factory function returns. But we set that up and name things correctly.

**Review by jsumners — DISMISSED** [review]

File: signer.js (hunk 5 above)
Comment (Uzlopak) [reply signer.js:8]: Ah. Allrighty. Will do

Comment (jsumners) [reply signer.js:8]: Yes.

File: signer.js (hunk 4 above)
Comment (Uzlopak) [reply signer.js:81]: @mcollina 

That is not correct. For the exposed shorthands sign and unsign, they will call validateSecrets before calling the underlying sign and unsign functions of the Signer. So there is now way, that the dummy key will be used.

File: signer.js (hunk 5 above)
Comment (Uzlopak) [reply signer.js]: renamed SignerFactory to Signer. Waiting for your feedback ;)

File: signer.js [hunk 6]
```diff
@@ -1,32 +1,101 @@
 'use strict'
```
Comment (Uzlopak) [inline signer.js:2]: @mcollina 
I would like to modify the header like this.

```suggestion

// Inspired by node-cookie-signature
// https://github.com/tj/node-cookie-signature
//
// The MIT License
// Copyright (c) 2012–2022 LearnBoost <tj@learnboost.com> and other contributors;

```

File: signer.js (hunk 3 above)
Comment (Uzlopak) [reply signer.js]: I would rather not extract the cookie-signature bits in its own file, as it is imho now a "logical unit"

File: README.md [hunk 7]
```diff
@@ -235,12 +235,12 @@ test('Request requires signed cookie', async () => {
...
 
-with signerFactory
+with Signer
 
 ```js
-const { signerFactory } = require('@fastify/cookie');
+const { Signer } = require('@fastify/cookie');
 
-const signer = signerFactory('secret');
+const signer = new Signer('secret');
```
Comment (jsumners) [inline README.md:243]: This is a breaking change.

File: plugin.js [hunk 8]
```diff
@@ -115,10 +115,10 @@ fastifyCookie.fastifyCookie = fastifyCookie
 fastifyCookie.default = fastifyCookie
 module.exports = fastifyCookie
 
-fastifyCookie.fastifyCookie.signerFactory = signerFactory
+fastifyCookie.fastifyCookie.Signer = Signer
```
Comment (jsumners) [inline plugin.js:118]: Breaking change.

File: plugin.js [hunk 9]
```diff
@@ -115,10 +115,10 @@ fastifyCookie.fastifyCookie = fastifyCookie
...
 fastifyCookie.fastifyCookie.sign = sign
 fastifyCookie.fastifyCookie.unsign = unsign
 
-module.exports.signerFactory = signerFactory
+module.exports.Signer = Signer
```
Comment (jsumners) [inline plugin.js:122]: Breaking change.

**Comment by Uzlopak** [comment]
Is it an issue that it is a breaking change? I could still export Signer also as signerFactory, keeping it backwards compatible...

**Comment by climba03003** [comment]
> Is it an issue that it is a breaking change? I could still export Signer also as signerFactory, keeping it backwards compatible...

Yes, please. Reduce breaking changes when possible.
Please keep the name `signerFactory` as it already exposed in `7.2.0`.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by budarin** [comment]
I would like more description in the documentation on the proposed features

**Comment by Uzlopak** [comment]
Ok so far?

**Comment by budarin** [comment]
> Ok so far?

any docs?

**Comment by Uzlopak** [comment]
Well i am actually more interested about the integration with e.g. @fastify/session. Should we e.g. use then the shorthandle on fastify.signCookie and fastify.unsignCookie instead of using sign and unsign. 

Other than that i think documentation is only needed for the algorithm option.

File: signer.js (hunk 4 above)
Comment (mcollina) [reply signer.js]: It's still pretty bad to see and look, and it's likely to come bite us in the future. Please fix.

**Review by mcollina — CHANGES_REQUESTED** [review]
Explicitly marking it as changes needed: drop the dummy secret.

File: signer.js (hunk 4 above)
Comment (Uzlopak) [reply signer.js]: done

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by U-4-E-A** [comment]
`cookie-signature` still exists in dependencies but does not appear to be required in either the library code or the testing. Should this be removed?

**Pull request merged**